### PR TITLE
lps: restore secondary cores on wake up

### PR DESCRIPTION
**Pull request by bkokoszx**
Primary core disables all secondary cores in platform_pg_int_handler() function when going to D0ix state. During coming out from D0ix state (D0ix -> D0), we are not able to restore secondary cores properly. This PR add commits that allow us to enable LPS flow to be used in multicore scenario properly. This PR:
- adds secondary cores data writeback before disabling them;
- adds secondary cores restore during D0->D0ix transition (pm_runtime_get, enable idc interrupt and send power up idc message);
- adds separate booting path in secondary_core_init() function i.e. in case when we have basic firmware structures already allocated (i.e. schedulers, notifier  etc.), which means we are not during cold boot process (memory has not been powered off), we just register and enable required interrupts.

### CODE REVIEW

File: src/init/init.c [hunk 1]
```diff
@@ -70,6 +71,49 @@ static inline void lp_sram_unpack(void)
...
+static int secondary_core_restore(void)
+{
+	int err = 0;
+
+	trace_point(TRACE_BOOT_PLATFORM_IRQ);
+
+	/* initialize interrupts */
+	platform_interrupt_init();
+
+	/* restore schedulers i.e. regist and enable scheduler interrupts */
```
[the remark this answers is not in the document]
Comment (bkokoszx) [reply src/init/init.c]: Updated

File: src/arch/xtensa/include/arch/lib/cpu.h [hunk 2]
```diff
@@ -28,6 +28,8 @@ int arch_cpu_enabled_cores(void);
 
 int arch_cpu_restore_secondary_cores(void);
 
+int arch_cpu_writeback_secondary_cores(void);
+
```
Comment (lyakh) [inline src/arch/xtensa/include/arch/lib/cpu.h:32]: to commit description: sorry, I don't see where in `platform_pg_int_handler()` all secondary cores are disabled?

File: src/ipc/ipc3/handler.c [hunk 3]
```diff
@@ -705,6 +705,11 @@ static int ipc_pm_gate(uint32_t header)
 		pm_runtime_disable(PM_RUNTIME_DSP, PLATFORM_PRIMARY_CORE_ID);
 		cpu_restore_secondary_cores();
 	} else {
+		/* before we enable pm runtime and perform D0->D0ix flow
+		 * (primary core powers off secondary cores in ) we have to
```
Comment (lyakh) [inline src/ipc/ipc3/handler.c:709]: is there a function name missing after "in" above?

File: src/ipc/ipc3/handler.c [hunk 4]
```diff
@@ -705,6 +705,11 @@ static int ipc_pm_gate(uint32_t header)
...
+		 * writeback all secondary core data.
+		 */
+		cpu_writeback_secondary_cores();
```
Comment (lyakh) [inline src/ipc/ipc3/handler.c:712]: We have flows where the host powers off secondary DSP cores using `SOF_IPC_PM_CORE_ENABLE`. Is `SOF_IPC_PM_GATE` also supposed to be able to switch secondary cores off without `SOF_IPC_PM_CORE_ENABLE` being used? As far as I understand in audio context, `SOF_IPC_PM_GATE` would typically be sent if the only active stream is keyword detection, running on the primary core. In that case all other cores should already be powered off?

File: src/drivers/intel/cavs/idc.c [hunk 5]
```diff
@@ -243,6 +243,35 @@ int platform_idc_init(void)
...
+ * \brief Restores IDC interrupt.
+ */
+int platform_idc_restore(void)
+{
+	struct idc *idc = *idc_get();
+	int core = cpu_get_id();
+	int ret;
+
+	interrupt_disable(idc->irq, idc);
+	interrupt_unregister(idc->irq, idc);
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/drivers/intel/cavs/idc.c:256]: Not really, in this scenario core power down sequence is not involved

File: src/init/init.c [hunk 6]
```diff
@@ -70,6 +71,49 @@ static inline void lp_sram_unpack(void)
...
+
+	/* restore idc i.e. register and enable idc interrupts */
+	trace_point(TRACE_BOOT_PLATFORM_IDC);
+	err = idc_restore();
+	if (err < 0)
+		return err;
+
+	trace_point(TRACE_BOOT_PLATFORM);
+
+	while (1)
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/init/init.c:114]: yes

File: src/ipc/ipc3/handler.c (hunk 4 above)
Comment (abonislawski) [reply src/ipc/ipc3/handler.c:712]: If LPS is involved then all secondary cores will be put to sleep by FW independently.
Check issue #4382 for more info about such scenario

File: src/schedule/edf_schedule.c [hunk 7]
```diff
@@ -298,6 +298,25 @@ static void scheduler_free_edf(void *data)
...
 
+static void scheduler_restore_edf(void *data)
+{
+	struct edf_schedule_data *edf_sch = data;
+	uint32_t flags;
+
+	irq_local_disable(flags);
+
+	interrupt_disable(edf_sch->irq, edf_sch);
+	interrupt_unregister(edf_sch->irq, edf_sch);
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/schedule/edf_schedule.c:309]: In this scenario core power down and scheduler_free() will not be called.
There will be nothing to restore if we will call scheduler_free :)

File: src/arch/xtensa/include/arch/lib/cpu.h (hunk 2 above)
Comment (abonislawski) [reply src/arch/xtensa/include/arch/lib/cpu.h:32]: pm_runtime_put

File: src/arch/xtensa/lib/cpu.c [hunk 8]
```diff
@@ -189,3 +189,28 @@ void cpu_power_down_core(void)
 	while (1)
 		arch_wait_for_interrupt(0);
 }
+
+int arch_cpu_restore_secondary_cores(void)
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/arch/xtensa/lib/cpu.c:193]: This code will only restore cores which was active previously, such a mask will be used inside this function (it is not available globally)

File: src/init/init.c [hunk 9]
```diff
@@ -70,6 +71,49 @@ static inline void lp_sram_unpack(void)
...
+{
+	struct idc *idc = *idc_get();
+	struct task *task = *task_main_get();
+	struct notify *notifier = *arch_notify_get();
+	struct schedulers *schedulers = *arch_schedulers_get();
+
+	/* check whether basic core structures has been already allocated. If they
+	   are available in memory, it means that this is not cold boot and memory
+	   has not been powered off. */
+	if (!idc || !task || !notifier || !schedulers)
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/init/init.c:85]: Good idea but I think we should still somehow validate all critical values in memory to make sure it is still valid to restore

File: src/arch/xtensa/lib/cpu.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/arch/xtensa/lib/cpu.c:193]: Why can we not have a global ```SHARED``` flag, possible as part of sof context ? I'm assuming this would be valid since it would be in the uncache alias.

File: src/drivers/intel/cavs/idc.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/drivers/intel/cavs/idc.c:256]: ok, so how do we tell the core is is about to power down ? It looks like we enter a WFI and the the LPS HW powers off ?

File: src/platform/intel/cavs/lib/pm_runtime.c [hunk 10]
```diff
@@ -567,6 +570,30 @@ void platform_pm_runtime_enable(uint32_t context, uint32_t index)
 	}
 }
 
+void platform_pm_runtime_waiti_writeback_en(uint32_t index)
+{
+	struct pm_runtime_data *prd = pm_runtime_data_get();
+	struct cavs_pm_runtime_data *pprd = prd->platform_data;
+
+	pprd->wait_writeback_core_mask |= BIT(index);
```
Comment (lgirdwood) [inline src/platform/intel/cavs/lib/pm_runtime.c:578]: Maybe I've mentioned this in another patch, but dont we need to lock this mask ?

File: src/arch/xtensa/lib/cpu.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/arch/xtensa/lib/cpu.c:193]: No cache issues here (only core0 will access it). That flag already exists - active_cores_mask, its static in cpu.c (but can be used anywhere from cpu_enabled_cores() if needed), 
It is used inside arch_cpu_restore_secondary_cores() (in arch_cpu_is_core_enabled(id) checks).

We can take it from cpu.c as cpu_enabled_cores(), run cpu_restore_secondary_cores(cpu_enabled_cores()) and inside anyway run helper functions like arch_cpu_is_core_enabled() which have local access to active cores mask. All of this is just weird and unnecessary. Maybe the name is just misleading and it will be better to call it cpu_restore_active_secondary_cores or something?

File: src/drivers/intel/cavs/idc.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/drivers/intel/cavs/idc.c:256]: We wont, lps just powers off - only pm_runtime_put without full power down sequence.
Thats why we have added writeback idc in pm gate handler, secondary core must have a chance to writeback before lps kicks in

File: src/platform/intel/cavs/lib/pm_runtime.c (hunk 10 above)
Comment (abonislawski) [reply src/platform/intel/cavs/lib/pm_runtime.c:578]: I think we should

**Comment by keyonjie** [comment]
> @bkokoszx @keyonjie @ranj063 lets have a proper shutdown sequence, this PR should not have to shutdown and reinitialize subsystems in the bringup. Topic for Tuesday call.

yes, we need to do more tearing down before shutting down a secondary core, I drafted a PR to showcase what I meant in the call here:
https://github.com/thesofproject/sof/pull/4718

**Comment by bkokoszx** [comment]
@lgirdwood @keyonjie @abonislawski @lyakh 
I've just updated this PR. I've added possibility for idc_free() and schedule_free() to disable only interrupts. I've also added flag for ```cpu_power_down_core()``` function i.e. CPU_POWER_DOWN_MEMORY_ON, which indicates not to free memory resources (basically it disables interrupts and perform writeback/invalidate of data) - ```cpu_power_down_core(CPU_POWER_DOWN_MEMORY_ON)``` is invoked for secondary cores in ```platform_wait_for_interrupt()``` when we should make a preparation before going to d0ix state.

To sum up:
- before going to d0ix state secondary cores performs writeback and disables interrupts;
- during booting up (after d0ix) secondary cores just enable idc and scheduler interrupts.

**Comment by lgirdwood** [comment]
@bkokoszx  the DMIC CI failures look unrelated for this PM related PR ?

**Comment by bkokoszx** [comment]
@lgirdwood 
I am not sure whether TIMEOUT on CML_RVP_SDW in ``check-suspend-resume-with-playback-5.sh`` test case is releated. If test just pauses and resumes pipelines without any D0ix flow then it should not be related with above change.

BTW
I have made one little, additional change:

``` c
static int ipc_pm_gate(uint32_t header)
{
	struct sof_ipc_pm_gate pm_gate;

	IPC_COPY_CMD(pm_gate, ipc_get()->comp_data);

	tr_info(&ipc_tr, "ipc: pm gate flags 0x%x", pm_gate.flags);

	/* pause dma trace firstly if needed */
	if (pm_gate.flags & SOF_PM_NO_TRACE)
		trace_off();

	if (pm_gate.flags & SOF_PM_PPG) {
		pm_runtime_disable(PM_RUNTIME_DSP, PLATFORM_PRIMARY_CORE_ID);
#if (CONFIG_CAVS_LPS)
		cpu_restore_secondary_cores();
#endif
	} else {
		/* before we enable pm runtime and perform D0->D0ix flow
		 * (primary core powers off secondary cores in
		 * platform_pg_int_handler) we have to prepare all secondary
		 * cores data for powering off (disable interrupt, perform
		 * cache writeback).
		 */
#if (CONFIG_CAVS_LPS)
		cpu_secondary_cores_prepare_d0ix();
#endif
		pm_runtime_enable(PM_RUNTIME_DSP, PLATFORM_PRIMARY_CORE_ID);
	}

	/* resume dma trace if needed */
	if (!(pm_gate.flags & SOF_PM_NO_TRACE))
		trace_on();

	return 0;
}
```
I invoke ```cpu_restore_secondary_cores()``` and ```cpu_secondary_cores_prepare_d0ix()``` only in case when LPS is enabled.

**Comment by bkokoszx** [comment]
@lgirdwood  Do you require some other changes?

**Review by lgirdwood — APPROVED** [review]
